### PR TITLE
[Spark] Show clusterBy in DESCRIBE HISTORY for clustered tables

### LINKED ISSUE #2470: [Feature Request] Support DESCRIBE HISTORY for clustered table
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

### Overview

Uber issue: https://github.com/delta-io/delta/issues/1874

Currently, there's no way to tell if a CREATE TABLE or OPTIMIZE was run on a clustered table. We should add the clustering columns information in operationParameters for DESCRIBE HISTORY command.

### Willingness to contribute

The Delta Lake Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature?

- [x] Yes. I can contribute this feature independently.
- [ ] Yes. I would be willing to contribute this feature with guidance from the Delta Lake community.
- [ ] No. I cannot contribute this feature at this time.

### PR DESCRIPTION
**Pull request by zedtang**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Resolves #2470 

Support showing `clusterBy` in DESCRIBE HISTORY for clustered table. Currently it supports create/replace table operations. The `clusterBy` key will always be present, even if we are not creating a clustered table. This behavior is consistent with `partitionBy`.

## How was this patch tested?

New unit tests.
## Does this PR introduce _any_ user-facing changes?

Yes, it adds `clusterBy` to DESCRIBE HISTORY's output.

### CODE REVIEW

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala [hunk 1]
```diff
@@ -60,6 +61,33 @@ trait ClusteredTableTestUtilsBase extends SparkFunSuite with SharedSparkSession
...
+      table: String
+  ): Unit = {
+    val deltaLog = if (table.startsWith("tahoe.") || table.startsWith("delta.")) {
+      // Path based table.
+      DeltaLog.forTable(spark, table.drop(6).replace("`", ""))
+    } else {
+      DeltaLog.forTable(spark, TableIdentifier(table))
+    }
+    val clusteringColumns =
+      ClusteringColumnInfo.extractLogicalNames(deltaLog.unsafeVolatileSnapshot)
```
Comment (weiluo-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala:75]: Can this `deltaLog.unsafeVolatileSnapshot` be avoided if you do `DeltaLog.forTableWithSnapshot()` above?

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala [hunk 2]
```diff
@@ -60,6 +61,33 @@ trait ClusteredTableTestUtilsBase extends SparkFunSuite with SharedSparkSession
...
+    val lastEvent = deltaLog.history.getHistory(Some(1)).head
+    lastEvent.operation match {
+      case "CLUSTER BY" =>
+        // Operation is [[DeltaOperations.ClusterBy]] - ALTER TABLE CLUSTER BY
+        assert(
+          lastEvent.operationParameters("newClusteringColumns") === clusteringColumns.mkString(",")
```
Comment (weiluo-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala:81]: Does this still hold if you cluster by NONE?

**Review by weiluo-db — CHANGES_REQUESTED** [review]

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala [hunk 3]
```diff
@@ -294,14 +294,15 @@ object DeltaOperations {
...
+      asSelect: Boolean = false,
+      clusterBy: Option[Seq[String]] = None
   ) extends Operation("CREATE TABLE" + s"${if (asSelect) " AS SELECT" else ""}") {
     override val parameters: Map[String, Any] = Map(
       "isManaged" -> isManaged.toString,
       "description" -> Option(metadata.description),
       "partitionBy" -> JsonUtils.toJson(metadata.partitionColumns),
       "properties" -> JsonUtils.toJson(metadata.configuration)
-    )
+    ) ++ (clusterBy.map(CLUSTERING_PARAMETER_KEY -> JsonUtils.toJson(_)))
```
Comment (imback82) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:305]: Can this go inside the above `()`?

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala [hunk 4]
```diff
@@ -60,6 +61,33 @@ trait ClusteredTableTestUtilsBase extends SparkFunSuite with SharedSparkSession
     assert(expectedClusteringColumns == actualClusteringColumns)
   }
 
+  // Verify the operation parameters of the last history event contains `clusterBy`.
+  protected def verifyDescribeHistoryOperationParameters(
+      table: String
+  ): Unit = {
+    val deltaLog = if (table.startsWith("tahoe.") || table.startsWith("delta.")) {
```
Comment (imback82) [inline spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala:68]: is this a common pattern to validate path based?

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala (hunk 3 above)
Comment (ami7o) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:305]: Shouldn't this parameter only populated when clusterBy is specified?

Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:305]: Hi @ami7o , when `clusterBy` is None, the right side of `++` will become `None` so it became a no-op.

Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:305]: Hi @imback82 , the only way I could think of is to do:
```
    override val parameters: Map[String, Any] = (Seq(
      "isManaged" -> isManaged.toString,
      "description" -> Option(metadata.description),
      "partitionBy" -> JsonUtils.toJson(metadata.partitionColumns),
      "properties" -> JsonUtils.toJson(metadata.configuration)) ++
      (clusterBy.map(CLUSTERING_PARAMETER_KEY -> JsonUtils.toJson(_)))).toMap
```

But I don't particularly like it. Let me know if you prefer this.

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala (hunk 4 above)
Comment (zedtang) [reply spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala]: Yeah, I think so

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala (hunk 1 above)
Comment (zedtang) [reply spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala]: good idea, updated!

File: spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala (hunk 2 above)
Comment (zedtang) [reply spark/src/test/scala/org/apache/spark/sql/delta/skipping/ClusteredTableTestUtils.scala:80]: Yes, in that case, `clusteringColumns` will be an empty list.


We have a test that exercises this path: https://github.com/delta-io/delta/blob/49f2625423e16a7d0f6c7892145c7b3eed329052/spark/src/test/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableDDLSuite.scala#L563

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala (hunk 3 above)
Comment (imback82) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:305]: I think we should be consistent with `paritionBy` when there is no `PARTITIONED BY` clause?

**Review by imback82 — COMMENTED** [review]
one question, but lgtm

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala (hunk 3 above)
Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:306]: Good idea. I updated the PR to use this new behavior: always have `clusterBy` even if there's no CLUSTER BY clause.

Comment (weiluo-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:306]: It seems that you forgot to remove `++ (clusterBy.map(CLUSTERING_PARAMETER_KEY -> JsonUtils.toJson(_)))` at line 306.

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala [hunk 5]
```diff
@@ -324,6 +325,7 @@ object DeltaOperations {
       "isManaged" -> isManaged.toString,
       "description" -> Option(metadata.description),
       "partitionBy" -> JsonUtils.toJson(metadata.partitionColumns),
+      CLUSTERING_PARAMETER_KEY -> JsonUtils.toJson(clusterBy.getOrElse(Seq.empty)),
       "properties" -> JsonUtils.toJson(metadata.configuration)
     ) ++ (clusterBy.map(CLUSTERING_PARAMETER_KEY -> JsonUtils.toJson(_)))
```
Comment (weiluo-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:330]: Ditto.

Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala]: thanks for catching it. Updated.

**Review by imback82 — APPROVED** [review]
LGTM!

**Review by weiluo-db — APPROVED** [review]

**Review by ami7o — APPROVED** [review]

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala [hunk 6]
```diff
@@ -491,17 +495,24 @@ object DeltaOperations {
...
+      clusterBy: Option[Seq[String]] = None
   ) extends OptimizeOrReorg(OPTIMIZE_OPERATION_NAME, predicate) {
     override val parameters: Map[String, Any] = super.parameters ++ Map(
       ZORDER_PARAMETER_KEY -> JsonUtils.toJson(zOrderBy),
+      CLUSTERING_PARAMETER_KEY -> JsonUtils.toJson(clusterBy.getOrElse(Seq.empty)),
       AUTO_COMPACTION_PARAMETER_KEY -> auto
-    )
+    ) ++ (if (clusterBy.nonEmpty) {
+      // When clustering columns are specified, we will clear the zOrderBy key.
+      Map(ZORDER_PARAMETER_KEY -> JsonUtils.toJson(Seq.empty))
```
Comment (dabao521) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:514]: Given this is going to be OSSed, can we just fold this logic into the above like
```
override val parameters: Map[String, Any] = super.parameters ++ Map(
      ZORDER_PARAMETER_KEY -> if (clusterBy.nonEmpty) JsonUtils.toJson(Seq.empty) else JsonUtils.toJson(zOrderBy),
```

Comment (dabao521) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:514]: Can we just fold this logic into the above like
```
override val parameters: Map[String, Any] = super.parameters ++ Map(
      ZORDER_PARAMETER_KEY -> if (clusterBy.nonEmpty) JsonUtils.toJson(Seq.empty) else JsonUtils.toJson(zOrderBy),
```

**Review by dabao521 — APPROVED** [review]

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala (hunk 6 above)
Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala]: Good suggestion, updated

**Review by tdas — APPROVED** [review]

**Pull request merged**